parser/worker.go

```diff
@@ -262,44 +262,67 @@ func (w Worker) ExportCommit(commit *tmtypes.Commit, vals *tmctypes.ResultValida
 	return nil
 }
 
-// ExportTxs accepts a slice of transactions and persists then inside the database.
+// saveTx accepts the transaction and persists it inside the database.
 // An error is returned if the write fails.
-func (w Worker) ExportTxs(txs []*types.Tx) error {
-	// Handle all the transactions inside the block
-	for _, tx := range txs {
-		// Save the transaction itself
-		err := w.db.SaveTx(tx)
+func (w Worker) saveTx(tx *types.Tx) error {
+	err := w.db.SaveTx(tx)
+	if err != nil {
+		return fmt.Errorf("failed to handle transaction with hash %s: %s", tx.TxHash, err)
+	}
+	return nil
+}
+
+// handleTx accepts the transaction and calls the tx handlers.
+func (w Worker) handleTx(tx *types.Tx) {
+	// Call the tx handlers
+	for _, module := range w.modules {
+		if transactionModule, ok := module.(modules.TransactionModule); ok {
+			err := transactionModule.HandleTx(tx)
+			if err != nil {
+				w.logger.TxError(module, tx, err)
+			}
+		}
+	}
+}
+
+// handleMessages accepts the transaction and handles messages contained
+// inside the transaction. An error is returned if the message unpacking
+// or calling handlers fails.
+func (w Worker) handleMessages(tx *types.Tx) error {
+	// Handle all the messages contained inside the transaction
+	for i, msg := range tx.Body.Messages {
+		var stdMsg sdk.Msg
+		err := w.codec.UnpackAny(msg, &stdMsg)
 		if err != nil {
-			return fmt.Errorf("failed to handle transaction with hash %s: %s", tx.TxHash, err)
+			return fmt.Errorf("error while unpacking message: %s", err)
 		}
 
-		// Call the tx handlers
+		// Call the handlers
 		for _, module := range w.modules {
-			if transactionModule, ok := module.(modules.TransactionModule); ok {
-				err = transactionModule.HandleTx(tx)
+			if messageModule, ok := module.(modules.MessageModule); ok {
+				err = messageModule.HandleMsg(i, stdMsg, tx)
 				if err != nil {
-					w.logger.TxError(module, tx, err)
+					w.logger.MsgError(module, tx, stdMsg, err)
 				}
 			}
 		}
+	}
+	return nil
+}
 
-		// Handle all the messages contained inside the transaction
-		for i, msg := range tx.Body.Messages {
-			var stdMsg sdk.Msg
-			err = w.codec.UnpackAny(msg, &stdMsg)
-			if err != nil {
-				return fmt.Errorf("error while unpacking message: %s", err)
-			}
+// ExportTxs accepts a slice of transactions and persists then inside the database.
+// An error is returned if the write fails.
+func (w Worker) ExportTxs(txs []*types.Tx) error {
+	for _, tx := range txs {
+		go w.handleTx(tx)
```

Comment:
This should be called **after** the `w.saveTx` as the modules might depend on the stored transaction to handle their messages (maybe a cross-table reference)